includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php

```diff
@@ -0,0 +1,143 @@
+use WP_Error;
+use WP_REST_Request;
+use WP_REST_Response;
+use WP_REST_Server;
+
+/**
+ * Class for handling rest routes for Key Metrics settings.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+class REST_Key_Metrics_Controller {
+
+	/**
+	 * Key_Metrics_Settings instance.
+	 *
+	 * @since n.e.x.t
+	 * @var Key_Metrics_Settings
+	 */
+	protected $settings;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Key_Metrics_Settings $settings Key Metrics settings.
+	 */
+	public function __construct( Key_Metrics_Settings $settings ) {
+		$this->settings = $settings;
+	}
+
+	/**
+	 * Registers functionality through WordPress hooks.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function register() {
+		add_filter(
+			'googlesitekit_rest_routes',
+			function ( $routes ) {
+				return array_merge( $routes, $this->get_rest_routes() );
+			}
+		);
+
+		add_filter(
+			'googlesitekit_apifetch_preload_paths',
+			function ( $paths ) {
+				return array_merge(
+					$paths,
+					array(
+						'/' . REST_Routes::REST_ROOT . '/core/user/data/key-metrics',
+					)
+				);
+			}
+		);
+	}
+
+	/**
+	 * Gets REST route instances.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return REST_Route[] List of REST_Route objects.
+	 */
+	protected function get_rest_routes() {
+		$has_capabilities = function() {
+			return current_user_can( Permissions::VIEW_SPLASH ) || current_user_can( Permissions::VIEW_DASHBOARD );
+		};
+
+		return array(
+			new REST_Route(
+				'core/user/data/key-metrics',
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => function () {
+						return new WP_REST_Response( $this->settings->get() );
+					},
+					'permission_callback' => $has_capabilities,
+				)
+			),
+			new REST_Route(
+				'core/user/data/key-metrics',
+				array(
+					'methods'             => WP_REST_Server::CREATABLE,
+					'callback'            => function ( WP_REST_Request $request ) {
+						// Data is already validated because we've defined the detailed schema.
+						// If the incoming data param doesn't match the schema, then WordPress
+						// will automatically return the rest_invalid_param error and we will
+						// never get to here.
+						$data = $request->get_param( 'data' );
+						$this->settings->merge( $data['settings'] );
+
+						return new WP_REST_Response( $this->settings->get() );
+					},
+					'permission_callback' => $has_capabilities,
+					'args'                => array(
+						'data' => array(
+							'type'       => 'object',
+							'required'   => true,
+							'properties' => array(
+								'settings' => array(
+									'type'       => 'object',
+									'required'   => true,
+									'properties' => array(
+										'isWidgetHidden' => array(
+											'type'     => 'boolean',
+											'required' => false,
+										),
+										'widgetSlugs'    => array(
+											'type'     => 'array',
+											'required' => false,
+											'items'    => array(
```

Comment:
I've spotted something that was missed in the AC. From the [Design Doc](https://docs.google.com/document/d/1k8NOo2xEHUQAcgYG0TB5sUjGcPiC8SWIalEdd9G8-9g/edit#heading=h.th9qnr6nkrac):

> The `POST:key-metrics` endpoint should also return an error (which should be displayed to the user) if more than four metrics were submitted somehow.

It doesn't appear to have been captured in any of the other issues. This can easily be accomplished using [`maxItems`](https://json-schema.org/understanding-json-schema/reference/array.html#id12) here:

```suggestion
											'maxItems' => 4,
											'items'    => array(
```

It would also be useful to cover an another test.

I realise this isn't in the AC so if you feel it's scope creep we can always create a separate issue for it, but I think it would be nice to include here.